### PR TITLE
[refactor-wip](FS) Use golang-like generator to replace FS's list interface

### PR DESCRIPTION
**Pull request by ByteYue**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/io/fs/file_system.h [hunk 1]
```diff
@@ -64,6 +65,43 @@ struct FileInfo {
     bool is_file;
 };
 
+struct FsListGenerator {
```
Comment (gavinchou) [inline be/src/io/fs/file_system.h:68]: class?
and, an interface/base class should always have sufficient and clear comments for all the functions.

consider naming FsListGenerator -> FileListIterator ?

File: be/src/io/fs/file_system.h [hunk 2]
```diff
@@ -64,6 +65,43 @@ struct FileInfo {
...
+    FsListGenerator() = default;
+    virtual ~FsListGenerator() = default;
+
+    virtual Status init() = 0;
```
Comment (gavinchou) [inline be/src/io/fs/file_system.h:72]: what is init() for? subclass should be initialized when it is created, we dont need to call baseclass.init().

File: be/src/io/fs/file_system.h [hunk 3]
```diff
@@ -64,6 +65,43 @@ struct FileInfo {
...
+
+    Result<FileInfo> next() {
```
Comment (gavinchou) [inline be/src/io/fs/file_system.h:74]: just make it pure virtual and eliminate `generageNext()`,
calling it here is not a good design.

File: be/src/io/fs/file_system.h [hunk 4]
```diff
@@ -64,6 +65,43 @@ struct FileInfo {
...
+        return Status::OK();
+    }
+
+    virtual bool has_next() const = 0;
+
+private:
+    virtual void generateNext() = 0;
+
+protected:
+    FileInfo file_info;
```
Comment (gavinchou) [inline be/src/io/fs/file_system.h:99]: what's this for ? cache?

**Review by gavinchou — CHANGES_REQUESTED** [review]

**Comment by yiguolei** [comment]
Any benifits for this refactor?

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by hello-stephen** [comment]
run p0

**Comment by ByteYue** [comment]
run buildall

**Pull request closed without merging**